nixos/modules/services/continuous-integration/gitlab-runner.nix

```diff
@@ -1,160 +1,423 @@
+            ];
+            dockerDisableCache = true;
+            preBuildScript = pkgs.writeScript "setup-container" '''
+              mkdir -p -m 0755 /nix/var/log/nix/drvs
+              mkdir -p -m 0755 /nix/var/nix/gcroots
+              mkdir -p -m 0755 /nix/var/nix/profiles
+              mkdir -p -m 0755 /nix/var/nix/temproots
+              mkdir -p -m 0755 /nix/var/nix/userpool
+              mkdir -p -m 1777 /nix/var/nix/gcroots/per-user
+              mkdir -p -m 1777 /nix/var/nix/profiles/per-user
+              mkdir -p -m 0755 /nix/var/nix/profiles/per-user/root
+              mkdir -p -m 0700 "$HOME/.nix-defexpr"
 
-  };
+              . ''${pkgs.nix}/etc/profile.d/nix.sh
+
+              ''${pkgs.nix}/bin/nix-env -i ''${concatStringsSep " " (with pkgs; [ nix cacert git openssh ])}
 
+              ''${pkgs.nix}/bin/nix-channel --add https://nixos.org/channels/nixpkgs-unstable
+              ''${pkgs.nix}/bin/nix-channel --update nixpkgs
+            ''';
+            environmentVariables = [
+              "ENV=/etc/profile"
+              "USER=root"
+              "NIX_REMOTE=daemon"
+              "PATH=/nix/var/nix/profiles/default/bin:/nix/var/nix/profiles/default/sbin:/bin:/sbin:/usr/bin:/usr/sbin"
+              "NIX_SSL_CERT_FILE=/nix/var/nix/profiles/default/etc/ssl/certs/ca-bundle.crt"
+            ];
+            tagList = [ "nix" ];
+          };
+          # runner for building docker images
+          docker-images = {
+            registrationConfigFile = pkgs.writeText "gitlab-runner-docker-images-registration" '''
+              CI_SERVER_URL=https://gitlab.com/
+              REGISTRATION_TOKEN=<token2>
+            ''';
+            dockerImage = "docker:stable";
+            dockerVolumes = [
+              "/var/run/docker.sock:/var/run/docker.sock"
+            ];
+            tagList = [ "docker-images" ];
+          };
+          # runner for everything else
+          default = {
+            registrationConfigFile = pkgs.writeText "gitlab-runner-default-registration" '''
+              CI_SERVER_URL=https://gitlab.com/
+              REGISTRATION_TOKEN=<token3>
+            ''';
+            dockerImage = "debian:stable";
+          };
+        }
+      '';
+      type = types.attrsOf (types.submodule {
+        options = {
+          registrationConfigFile = mkOption {
+            description = ''
+              File with environment variables used for gitlab-runner registration.
+              A list of all supported environment variables can be found in
+              <literal>gitlab-runner register --help</literal>.
+
+              Ones that you probably want to set is
+
+              <literal>CI_SERVER_URL=&lt;CI server URL&gt;</literal>
+
+              <literal>REGISTRATION_TOKEN=&lt;registration secret&gt;</literal>
+            '';
+            type = types.path;
+          };
+          environmentVariables = mkOption {
```

Comment:
We usually allow passing a dict `environment` here, and render it through nix, or a `environmentFile` (if the environment contains secrets).

Reply:
Yeah, i can make this an `attrset`.
Secrets could be passed using `registrationConfigFile`.

Reply:
Done.